pkg/telemetrylogs/key_evolution_metadata.go

```diff
@@ -0,0 +1,163 @@
+
+	rows, err := store.ClickhouseDB().Query(ctx, query, args...)
+	if err != nil {
+		logger.WarnContext(ctx, "Failed to fetch key evolution metadata from ClickHouse", "error", err)
+		return
+	}
+	defer rows.Close()
+
+	// Group metadata by base_column
+	metadataByKey := make(map[string][]*telemetrytypes.KeyEvolutionMetadataKey)
+
+	for rows.Next() {
+		var (
+			baseColumn     string
+			baseColumnType string
+			newColumn      string
+			newColumnType  string
+			releaseTime    uint64
+		)
+
+		if err := rows.Scan(&baseColumn, &baseColumnType, &newColumn, &newColumnType, &releaseTime); err != nil {
+			logger.WarnContext(ctx, "Failed to scan key evolution metadata row", "error", err)
+			continue
+		}
+
+		key := &telemetrytypes.KeyEvolutionMetadataKey{
+			BaseColumn:     baseColumn,
+			BaseColumnType: baseColumnType,
+			NewColumn:      newColumn,
+			NewColumnType:  newColumnType,
+			ReleaseTime:    time.Unix(0, int64(releaseTime)),
+		}
+
+		metadataByKey[baseColumn] = append(metadataByKey[baseColumn], key)
+	}
+
+	if err := rows.Err(); err != nil {
+		logger.WarnContext(ctx, "Error iterating key evolution metadata rows", "error", err)
+		return
+	}
+
+	// Store each key's metadata in cache
+	for keyName, keys := range metadataByKey {
+		cacheKey := KeyEvolutionMetadataCacheKeyPrefix + keyName
+		cachedData := &CachedKeyEvolutionMetadata{Keys: keys}
+		if err := k.cache.Set(ctx, orgID, cacheKey, cachedData, 24*time.Hour); err != nil {
+			logger.WarnContext(ctx, "Failed to set key evolution metadata in cache", "key", keyName, "error", err)
+		}
+	}
+
+	logger.DebugContext(ctx, "Successfully fetched key evolution metadata from ClickHouse", "count", len(metadataByKey))
+}
+
+// Add adds a metadata key for the given key name and orgId.
+// This is primarily for testing purposes. In production, data should come from ClickHouse.
+func (k *KeyEvolutionMetadata) Add(ctx context.Context, orgId valuer.UUID, keyName string, key *telemetrytypes.KeyEvolutionMetadataKey) {
+	k.logger.WarnContext(ctx, "Add is not implemented for key evolution metadata")
+
+}
+
+// Get retrieves all metadata keys for the given key name and orgId from cache.
+// Returns an empty slice if the key is not found in cache.
+func (k *KeyEvolutionMetadata) Get(ctx context.Context, orgId valuer.UUID, keyName string) []*telemetrytypes.KeyEvolutionMetadataKey {
+
+	cacheKey := KeyEvolutionMetadataCacheKeyPrefix + keyName
+	var cachedData CachedKeyEvolutionMetadata
+	if err := k.cache.Get(ctx, orgId, cacheKey, &cachedData); err != nil {
+		// Cache miss - fetch from ClickHouse and try again
+		k.fetchFromClickHouse(ctx, orgId)
+
+		// Check cache again after fetching
+		if err := k.cache.Get(ctx, orgId, cacheKey, &cachedData); err != nil {
+			return nil
+		}
+	}
+
+	// Return a copy to prevent external modification
+	result := make([]*telemetrytypes.KeyEvolutionMetadataKey, len(cachedData.Keys))
+	copy(result, cachedData.Keys)
```

Comment:
Is this really needed? Aren't new objects created with each `Get` because the data is JSON serialised/deserialised?